File: account/src/main/java/serp/project/account/ui/controller/PermissionController.java
/**
 * Author: QuanTuanHuy
 * Description: Part of Serp Project
 */

package serp.project.account.ui.controller;

import jakarta.validation.Valid;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import lombok.RequiredArgsConstructor;
import serp.project.account.core.domain.dto.request.CreatePermissionDto;
import serp.project.account.core.usecase.PermissionUseCase;

import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.GetMapping;

@RestController
@RequiredArgsConstructor
@RequestMapping("/api/v1/permissions")
public class PermissionController {
    private final PermissionUseCase permissionUseCase;

    @PostMapping
    public ResponseEntity<?> createPermission(@Valid @RequestBody CreatePermissionDto request) {
        var response = permissionUseCase.createPermission(request);
        return ResponseEntity.status(response.getCode()).body(response);
    }

    @GetMapping
    public ResponseEntity<?> getAllPermissions() {
        var response = permissionUseCase.getAllPermissions();
        return ResponseEntity.status(response.getCode()).body(response);
    }

}

File: api_gateway/src/ui/controller/account/permission_controller.go
/*
Author: QuanTuanHuy
Description: Part of Serp Project
*/

package controller

import (
	"github.com/gin-gonic/gin"
	"github.com/serp/api-gateway/src/core/domain/constant"
	request "github.com/serp/api-gateway/src/core/domain/dto/request/account"
	service "github.com/serp/api-gateway/src/core/service/account"
	"github.com/serp/api-gateway/src/kernel/utils"
)

type PermissionController struct {
	permissionService service.IPermissionService
}

func (p *PermissionController) CreatePermission(c *gin.Context) {
	var req request.CreatePermissionDto
	if err := c.ShouldBindJSON(&req); err != nil {
		utils.AbortErrorHandle(c, constant.GeneralBadRequest)
		return
	}

	res, err := p.permissionService.CreatePermission(c.Request.Context(), &req)
	if err != nil {
		utils.AbortErrorHandle(c, constant.GeneralInternalServerError)
		return
	}
	c.JSON(res.Code, res)
}

func (p *PermissionController) GetAllPermissions(c *gin.Context) {
	res, err := p.permissionService.GetAllPermissions(c.Request.Context())
	if err != nil {
		utils.AbortErrorHandle(c, constant.GeneralInternalServerError)
		return
	}
	c.JSON(res.Code, res)
}

func NewPermissionController(permissionService service.IPermissionService) *PermissionController {
	return &PermissionController{
		permissionService: permissionService,
	}
}
